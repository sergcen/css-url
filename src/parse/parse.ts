import * as URL from 'url';

const URL_PATTERNS = [
    /(url\(\s*['"]?)([^"')]+)(["']?\s*\))/g,
    /(AlphaImageLoader\(\s*src=['"]?)([^"')]+)(["'])/g,
];

const getPatterns = (str: string) =>
    URL_PATTERNS.filter((pattern) => pattern.test(str));

export type TReplacerFN = (url: string, before: string, after: string) => string

export const parseAndReplace: (str: string, replacer?: TReplacerFN) => { parsedUrls: URL.Url[], resultUrls: URL.Url[], result: string } = (str, replacer) => {
    const patterns = getPatterns(str);
    const parsedUrls: string[] = [];
    const resultUrls: string[] = [];

    for (let pattern of patterns) {
        str = str.replace(pattern, (_, before, url, after) => {
            parsedUrls.push(url);

            if (replacer) {
                const replaced = replacer(url, before, after);
                resultUrls.push(replaced);

                return before + replaced + after;
            }
            return '';
        });
    }

    return {
        result: str,
        // https://github.com/nodejs/node/issues/12682
        resultUrls: resultUrls.map(url => URL.parse(url)),
        // https://github.com/nodejs/node/issues/12682
        parsedUrls: parsedUrls.map(url => URL.parse(url))
    }
};

export const parse: (str: string) => URL.Url[] = (str) => {
    return parseAndReplace(str).parsedUrls;
};

export const replace: (str: string, replacer: TReplacerFN) => string = (str, replacer) => {
    return parseAndReplace(str, replacer).result;
}
