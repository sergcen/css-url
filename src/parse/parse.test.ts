import { parse, replace, parseAndReplace } from './parse';

const urls = [
    'url("one")',
    `url('two')`,
    `url(three)`,
    `url(/five)`,
    'url(#six)',
    'url(./some/path/1.svg)',
    'url(./some/path/1.png)',
    'url(./some/path/1.png#some-hash)',
    'url(./some/path/1.png?query=1)',
    `url('some/path/1.png?query=2#some-hash')`,
    `-filter:progid:DXImageTransform.Microsoft.AlphaImageLoader(src='path/1.gif',sizingMethod=scale)`
]
const multiUrls = `${urls.join(', ')};`;

describe('parse css url', () => {
    it('parse base', () => {
        expect(parse(multiUrls)).toMatchSnapshot();
    });

    it('replace base', () => {
        expect(replace(multiUrls, (url) => url.replace('path', 'replaced-path'))).toMatchSnapshot();
    });

    it('parse and replace', () => {
        expect(parseAndReplace(multiUrls, (url) => url.replace('path', 'replaced-path'))).toMatchSnapshot();
    });
})
