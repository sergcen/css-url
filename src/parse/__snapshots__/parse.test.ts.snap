// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse css url parse and replace 1`] = `
Object {
  "parsedUrls": Array [
    Url {
      "auth": null,
      "hash": null,
      "host": null,
      "hostname": null,
      "href": "one",
      "path": "one",
      "pathname": "one",
      "port": null,
      "protocol": null,
      "query": null,
      "search": null,
      "slashes": null,
    },
    Url {
      "auth": null,
      "hash": null,
      "host": null,
      "hostname": null,
      "href": "two",
      "path": "two",
      "pathname": "two",
      "port": null,
      "protocol": null,
      "query": null,
      "search": null,
      "slashes": null,
    },
    Url {
      "auth": null,
      "hash": null,
      "host": null,
      "hostname": null,
      "href": "three",
      "path": "three",
      "pathname": "three",
      "port": null,
      "protocol": null,
      "query": null,
      "search": null,
      "slashes": null,
    },
    Url {
      "auth": null,
      "hash": null,
      "host": null,
      "hostname": null,
      "href": "/five",
      "path": "/five",
      "pathname": "/five",
      "port": null,
      "protocol": null,
      "query": null,
      "search": null,
      "slashes": null,
    },
    Url {
      "auth": null,
      "hash": "#six",
      "host": null,
      "hostname": null,
      "href": "#six",
      "path": null,
      "pathname": null,
      "port": null,
      "protocol": null,
      "query": null,
      "search": null,
      "slashes": null,
    },
    Url {
      "auth": null,
      "hash": null,
      "host": null,
      "hostname": null,
      "href": "./some/path/1.svg",
      "path": "./some/path/1.svg",
      "pathname": "./some/path/1.svg",
      "port": null,
      "protocol": null,
      "query": null,
      "search": null,
      "slashes": null,
    },
    Url {
      "auth": null,
      "hash": null,
      "host": null,
      "hostname": null,
      "href": "./some/path/1.png",
      "path": "./some/path/1.png",
      "pathname": "./some/path/1.png",
      "port": null,
      "protocol": null,
      "query": null,
      "search": null,
      "slashes": null,
    },
    Url {
      "auth": null,
      "hash": "#some-hash",
      "host": null,
      "hostname": null,
      "href": "./some/path/1.png#some-hash",
      "path": "./some/path/1.png",
      "pathname": "./some/path/1.png",
      "port": null,
      "protocol": null,
      "query": null,
      "search": null,
      "slashes": null,
    },
    Url {
      "auth": null,
      "hash": null,
      "host": null,
      "hostname": null,
      "href": "./some/path/1.png?query=1",
      "path": "./some/path/1.png?query=1",
      "pathname": "./some/path/1.png",
      "port": null,
      "protocol": null,
      "query": "query=1",
      "search": "?query=1",
      "slashes": null,
    },
    Url {
      "auth": null,
      "hash": "#some-hash",
      "host": null,
      "hostname": null,
      "href": "some/path/1.png?query=2#some-hash",
      "path": "some/path/1.png?query=2",
      "pathname": "some/path/1.png",
      "port": null,
      "protocol": null,
      "query": "query=2",
      "search": "?query=2",
      "slashes": null,
    },
    Url {
      "auth": null,
      "hash": null,
      "host": null,
      "hostname": null,
      "href": "path/1.gif",
      "path": "path/1.gif",
      "pathname": "path/1.gif",
      "port": null,
      "protocol": null,
      "query": null,
      "search": null,
      "slashes": null,
    },
  ],
  "result": "url(\\"one\\"), url('two'), url(three), url(/five), url(#six), url(./some/replaced-path/1.svg), url(./some/replaced-path/1.png), url(./some/replaced-path/1.png#some-hash), url(./some/replaced-path/1.png?query=1), url('some/replaced-path/1.png?query=2#some-hash'), -filter:progid:DXImageTransform.Microsoft.AlphaImageLoader(src='replaced-path/1.gif',sizingMethod=scale);",
  "resultUrls": Array [
    Url {
      "auth": null,
      "hash": null,
      "host": null,
      "hostname": null,
      "href": "one",
      "path": "one",
      "pathname": "one",
      "port": null,
      "protocol": null,
      "query": null,
      "search": null,
      "slashes": null,
    },
    Url {
      "auth": null,
      "hash": null,
      "host": null,
      "hostname": null,
      "href": "two",
      "path": "two",
      "pathname": "two",
      "port": null,
      "protocol": null,
      "query": null,
      "search": null,
      "slashes": null,
    },
    Url {
      "auth": null,
      "hash": null,
      "host": null,
      "hostname": null,
      "href": "three",
      "path": "three",
      "pathname": "three",
      "port": null,
      "protocol": null,
      "query": null,
      "search": null,
      "slashes": null,
    },
    Url {
      "auth": null,
      "hash": null,
      "host": null,
      "hostname": null,
      "href": "/five",
      "path": "/five",
      "pathname": "/five",
      "port": null,
      "protocol": null,
      "query": null,
      "search": null,
      "slashes": null,
    },
    Url {
      "auth": null,
      "hash": "#six",
      "host": null,
      "hostname": null,
      "href": "#six",
      "path": null,
      "pathname": null,
      "port": null,
      "protocol": null,
      "query": null,
      "search": null,
      "slashes": null,
    },
    Url {
      "auth": null,
      "hash": null,
      "host": null,
      "hostname": null,
      "href": "./some/replaced-path/1.svg",
      "path": "./some/replaced-path/1.svg",
      "pathname": "./some/replaced-path/1.svg",
      "port": null,
      "protocol": null,
      "query": null,
      "search": null,
      "slashes": null,
    },
    Url {
      "auth": null,
      "hash": null,
      "host": null,
      "hostname": null,
      "href": "./some/replaced-path/1.png",
      "path": "./some/replaced-path/1.png",
      "pathname": "./some/replaced-path/1.png",
      "port": null,
      "protocol": null,
      "query": null,
      "search": null,
      "slashes": null,
    },
    Url {
      "auth": null,
      "hash": "#some-hash",
      "host": null,
      "hostname": null,
      "href": "./some/replaced-path/1.png#some-hash",
      "path": "./some/replaced-path/1.png",
      "pathname": "./some/replaced-path/1.png",
      "port": null,
      "protocol": null,
      "query": null,
      "search": null,
      "slashes": null,
    },
    Url {
      "auth": null,
      "hash": null,
      "host": null,
      "hostname": null,
      "href": "./some/replaced-path/1.png?query=1",
      "path": "./some/replaced-path/1.png?query=1",
      "pathname": "./some/replaced-path/1.png",
      "port": null,
      "protocol": null,
      "query": "query=1",
      "search": "?query=1",
      "slashes": null,
    },
    Url {
      "auth": null,
      "hash": "#some-hash",
      "host": null,
      "hostname": null,
      "href": "some/replaced-path/1.png?query=2#some-hash",
      "path": "some/replaced-path/1.png?query=2",
      "pathname": "some/replaced-path/1.png",
      "port": null,
      "protocol": null,
      "query": "query=2",
      "search": "?query=2",
      "slashes": null,
    },
    Url {
      "auth": null,
      "hash": null,
      "host": null,
      "hostname": null,
      "href": "replaced-path/1.gif",
      "path": "replaced-path/1.gif",
      "pathname": "replaced-path/1.gif",
      "port": null,
      "protocol": null,
      "query": null,
      "search": null,
      "slashes": null,
    },
  ],
}
`;

exports[`parse css url parse base 1`] = `
Array [
  Url {
    "auth": null,
    "hash": null,
    "host": null,
    "hostname": null,
    "href": "one",
    "path": "one",
    "pathname": "one",
    "port": null,
    "protocol": null,
    "query": null,
    "search": null,
    "slashes": null,
  },
  Url {
    "auth": null,
    "hash": null,
    "host": null,
    "hostname": null,
    "href": "two",
    "path": "two",
    "pathname": "two",
    "port": null,
    "protocol": null,
    "query": null,
    "search": null,
    "slashes": null,
  },
  Url {
    "auth": null,
    "hash": null,
    "host": null,
    "hostname": null,
    "href": "three",
    "path": "three",
    "pathname": "three",
    "port": null,
    "protocol": null,
    "query": null,
    "search": null,
    "slashes": null,
  },
  Url {
    "auth": null,
    "hash": null,
    "host": null,
    "hostname": null,
    "href": "/five",
    "path": "/five",
    "pathname": "/five",
    "port": null,
    "protocol": null,
    "query": null,
    "search": null,
    "slashes": null,
  },
  Url {
    "auth": null,
    "hash": "#six",
    "host": null,
    "hostname": null,
    "href": "#six",
    "path": null,
    "pathname": null,
    "port": null,
    "protocol": null,
    "query": null,
    "search": null,
    "slashes": null,
  },
  Url {
    "auth": null,
    "hash": null,
    "host": null,
    "hostname": null,
    "href": "./some/path/1.svg",
    "path": "./some/path/1.svg",
    "pathname": "./some/path/1.svg",
    "port": null,
    "protocol": null,
    "query": null,
    "search": null,
    "slashes": null,
  },
  Url {
    "auth": null,
    "hash": null,
    "host": null,
    "hostname": null,
    "href": "./some/path/1.png",
    "path": "./some/path/1.png",
    "pathname": "./some/path/1.png",
    "port": null,
    "protocol": null,
    "query": null,
    "search": null,
    "slashes": null,
  },
  Url {
    "auth": null,
    "hash": "#some-hash",
    "host": null,
    "hostname": null,
    "href": "./some/path/1.png#some-hash",
    "path": "./some/path/1.png",
    "pathname": "./some/path/1.png",
    "port": null,
    "protocol": null,
    "query": null,
    "search": null,
    "slashes": null,
  },
  Url {
    "auth": null,
    "hash": null,
    "host": null,
    "hostname": null,
    "href": "./some/path/1.png?query=1",
    "path": "./some/path/1.png?query=1",
    "pathname": "./some/path/1.png",
    "port": null,
    "protocol": null,
    "query": "query=1",
    "search": "?query=1",
    "slashes": null,
  },
  Url {
    "auth": null,
    "hash": "#some-hash",
    "host": null,
    "hostname": null,
    "href": "some/path/1.png?query=2#some-hash",
    "path": "some/path/1.png?query=2",
    "pathname": "some/path/1.png",
    "port": null,
    "protocol": null,
    "query": "query=2",
    "search": "?query=2",
    "slashes": null,
  },
  Url {
    "auth": null,
    "hash": null,
    "host": null,
    "hostname": null,
    "href": "path/1.gif",
    "path": "path/1.gif",
    "pathname": "path/1.gif",
    "port": null,
    "protocol": null,
    "query": null,
    "search": null,
    "slashes": null,
  },
]
`;

exports[`parse css url replace base 1`] = `"url(\\"one\\"), url('two'), url(three), url(/five), url(#six), url(./some/replaced-path/1.svg), url(./some/replaced-path/1.png), url(./some/replaced-path/1.png#some-hash), url(./some/replaced-path/1.png?query=1), url('some/replaced-path/1.png?query=2#some-hash'), -filter:progid:DXImageTransform.Microsoft.AlphaImageLoader(src='replaced-path/1.gif',sizingMethod=scale);"`;
